File: src/game/ts/players/avoid_others.ts
import { Playable } from "../runner";
import { createPlayer, toXY } from "./common";

const singleton: Playable = {
	play({ width, availableCells, neighbors, mass }) {
		const choice = [...availableCells()]
			.map(pos => ({
				pos,
				neighborMass: neighbors(pos)
					.map(mass)
					.reduce((acc, x) => acc + x, 0),
			}))
			.sort((a, b) => a.neighborMass - b.neighborMass);

		return toXY(choice[0].pos, width);
	},
};

export const AvoidOthers = createPlayer(singleton, {
	id: "AvoidOthers",
	wasm: true,
	name: "Avoid others",
	description: "Avoids placing cells in dense areas.",
});

File: src/game/rust/src/players/avoid_others.rs
use crate::{runner::GameContext, Player, Pos};

#[derive(Clone)]
pub struct AvoidOthers {}

impl Player for AvoidOthers {
    fn play(&mut self, ctx: GameContext) -> Result<Pos, &'static str> {
        let choice = ctx.available_cells();
        let mut choice: Vec<_> = choice
            .iter()
            .map(|pos| {
                let neighbor_mass_sum = ctx
                    .neighbors(pos)
                    .iter()
                    .map(|(n, _)| ctx.mass(&n))
                    .fold(0, |acc, x| acc + x);

                (pos, neighbor_mass_sum)
            })
            .collect();

        choice.sort_by(|a, b| a.1.cmp(&b.1));

        return Ok(*choice[0].0);
    }
}
